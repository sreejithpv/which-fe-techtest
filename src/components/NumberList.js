
import React, { useState, useCallback} from 'react';
import PropTypes from 'prop-types';

import { getNumberList } from '../utils';
import NumberBox from './NumberBox';

import '../styles/NumberList.css';

function NumberList({ maxNumberLimit }) {
  const [numberList, setNumberList] = useState(getNumberList({maxNumberLimit}));

  const selectNumber = (selectedNumber) => {
    const numList =getNumberList({
      maxNumberLimit,
      selectedNumber,
      numberList,
    })
    setNumberList([...numList]);
    };

  const onClick = useCallback((selectedNumber) => selectNumber(selectedNumber),[]);
  const onKeyPress = useCallback((selectedNumber, e) => { 
    if(e.keyCode === 13) { 
      selectNumber(selectedNumber) 
    }
    },[]);

  return (
    <ul className="grid-container">
      {numberList.map(numberItem => (
        <NumberBox 
          key={numberItem.value}
          onClick={() => !numberItem.isSelected ? onClick(numberItem.value): false}
          onKeyPress={(e) => !numberItem.isSelected ? onKeyPress(numberItem.value, e): false}
          value={numberItem.value}
          isSelected={numberItem.isSelected}
          isMultipliable={numberItem.isMultipliable}
        />
      ))}
    </ul>
  );
}

NumberList.propTypes = {
  maxNumberLimit: PropTypes.number,
};

NumberList.defaultProps = {
  maxNumberLimit: 12,
}

export default NumberList;
