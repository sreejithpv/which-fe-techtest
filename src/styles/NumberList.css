.grid-container {
    display: grid;
    grid-template-columns: auto auto auto;
    background-color: #118df3;
    padding: 10px;
    list-style: none;
  }
  .grid-item {
    background-color: rgba(255, 255, 255, 0.8);
    border: 1px solid rgba(0, 0, 0, 0.8);
    padding: 20%;
    font-size: 5em;
    text-align: center;
    cursor: pointer;
    transition: background-color 1s ease;
    display: flex;
    justify-content: center;
    align-items: center;
  }
  .grid-item.active {
    background-color: rgba(180, 243, 185, 0.8);
  }

  .grid-item.selected {
    background-color: rgba(14, 189, 212, 0.8);
    cursor: auto;
  }

  @media only screen and (max-width: 768px) {
    .grid-container {
        grid-template-columns: auto auto;
    }
  }
  @media only screen and (max-width: 576px) {
    .grid-container {
        grid-template-columns: auto;
    }
  }