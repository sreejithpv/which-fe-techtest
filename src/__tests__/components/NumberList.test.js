import '@testing-library/jest-dom'
import TestRenderer from 'react-test-renderer';
import { screen, render, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import NumberList from '../../components/NumberList';

const baseProps = {
  maxNumberLimit: 4,
}
const renderFun = ( props = baseProps ) => TestRenderer.create(<NumberList {...props} />);
const keyEvent = { key: 'Enter', keyCode: 13 };

describe('NumberList Component', () => {
  it("should render snapshot with default props", () => {
    const tree = renderFun();
    expect(tree.toJSON()).toMatchSnapshot();
  });
  describe('Events', () => {
    it("should trigger click", () => {
      render(<NumberList {...baseProps} />);
      const element = screen.getByText('2');
      expect(element).toBeInTheDocument();
      userEvent.click(element);
      expect(element.classList.contains('selected')).toBe(true)
    });

    it("should not multiple trigger click on selected element", () => {
      render(<NumberList {...baseProps} />);
      const element = screen.getByText('2');
      expect(element).toBeInTheDocument();
      userEvent.click(element);
      expect(element.classList.contains('selected')).toBe(true);
      userEvent.click(element);
      expect(element.classList.contains('selected')).toBe(true);
    });

    it("should trigger keydown event", () => {
      render(<NumberList {...baseProps} />);
      const element = screen.getByText('2');
      expect(element).toBeInTheDocument();
      fireEvent.keyDown(element, {...keyEvent});
      expect(element.classList.contains('selected')).toBe(true);
    });

    it("should trigger keydown event based on keyCode", () => {
      render(<NumberList {...baseProps} />);
      const element = screen.getByText('2');
      expect(element).toBeInTheDocument();
      fireEvent.keyDown(element, { key: '', keyCode: 13 });
      expect(element.classList.contains('selected')).toBe(true);
    });

    it("should not multiple keydown event on selected element", () => {
      render(<NumberList {...baseProps} />);
      const element = screen.getByText('2');
      expect(element).toBeInTheDocument();
      fireEvent.keyDown(element, {...keyEvent})
      expect(element.classList.contains('selected')).toBe(true);
      fireEvent.keyDown(element, {...keyEvent});
      expect(element.classList.contains('selected')).toBe(true);
    });

    it("should add active on all the boxes which are isMultipliable of 8", () => {
      render(<NumberList maxNumberLimit={8} />);
      const element = screen.getByText('8');
      expect(element).toBeInTheDocument();
      userEvent.click(element);
      const elementWith2 = screen.getByText('2');
      const elementWith4 = screen.getByText('4');
      expect(elementWith2.classList.contains('active')).toBe(true);
      expect(elementWith4.classList.contains('active')).toBe(true);
    });
  });
});
