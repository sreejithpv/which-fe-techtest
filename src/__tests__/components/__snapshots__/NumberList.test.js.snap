// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberList Component should render snapshot with default props 1`] = `
<ul
  className="grid-container"
>
  <li
    className="grid-item  "
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="0"
  >
    1
  </li>
  <li
    className="grid-item  "
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="0"
  >
    2
  </li>
  <li
    className="grid-item  "
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="0"
  >
    3
  </li>
  <li
    className="grid-item  "
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="0"
  >
    4
  </li>
</ul>
`;
